### PR TITLE
Processing data in models

### PR DESCRIPTION
**Pull request by Elerqsousy**
## In this milestone:

Created model classes for all entities as shown in the [ERD diagram](https://github.com/microverseinc/curriculum-rails/blob/main/blog-app/images/blog_app_erd_v1_1.png).
Set up associations between models.
Set up seeds files in db to handle the following: 
- Created two users 
- Created five posts
- Create six comments for the first post

Created the following custom methods:
- A method that returns the 3 most recent posts for a given user.
- A method that updates the posts counter for a user.
- A method which returns the 5 most recent comments for a given post.
- A method that updates the comments counter for a post.
- A method that updates the likes counter for a post.

Tested all methods using the seed file data and `rails c` command.

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,9 @@
+class User < ApplicationRecord
+  has_many :posts, foreign_key: 'author_id', dependent: :destroy
+  has_many :comments, foreign_key: 'user_id', dependent: :destroy
+  has_many :likes, foreign_key: 'user_id', dependent: :destroy
+
+  def recent_posts
+    Post.where(author_id: id).order(created_at: :desc).limit(3)
```
Comment (bienvenuushindi) [inline app/models/user.rb]: :orange_circle: [OPTIONAL] Below is a more concise and clear way to achieve the same result as the previous code: 
```
def recent_posts
    posts.order(created_at: :desc).limit(3)
 end
```

 This leverages ActiveRecord associations, using  the  has_many :posts association set up in your this model.

File: app/models/post.rb [hunk 2]
```diff
@@ -0,0 +1,19 @@
+class Post < ApplicationRecord
+  belongs_to :author, class_name: 'User'
+
+  has_many :comments, foreign_key: 'post_id', dependent: :destroy
+  has_many :likes, foreign_key: 'post_id', dependent: :destroy
+
+  after_save :update_user_posts_counter
+
+  def recent_comments
+    Comment.where(post_id: id).order(created_at: :desc).limit(5)
```
Comment (bienvenuushindi) [inline app/models/post.rb]: :orange_circle: [OPTIONAL] Below is a more concise and clear way to achieve the same result as the previous code: 
```
def recent_comments
    comments.order(created_at: :desc).limit(5)
 end
```

 This leverages ActiveRecord associations, using the  has_many :comments association set up in this model.

File: app/models/post.rb [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+  end
+
+  private
+
+  def update_user_posts_counter
+    user_posts_count = Post.where(author_id:).count
+    author.update(postsCounter: user_posts_count)
+  end
```
Comment (bienvenuushindi) [inline app/models/post.rb]: :orange_circle: [OPTIONAL] Below is a more concise and clear way to achieve the same result as the previous code: 

  def update_user_posts_counter
    author.increment!(:postsCounter)
  end

File: app/models/post.rb [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+  def recent_comments
+    Comment.where(post_id: id).order(created_at: :desc).limit(5)
+  end
+
+  private
+
+  def update_user_posts_counter
+    user_posts_count = Post.where(author_id:).count
+    author.update(postsCounter: user_posts_count)
```
Comment (bienvenuushindi) [inline app/models/post.rb]: :red_circle: [REQUIRED] Please ensure that you adhere to Ruby's best practices for variable naming. The naming convention for variables is` snake_case` for methods and variables. So `postsCounter` should be `posts_counter`

File: app/models/likes [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+    belongs_to :user
+    belongs_to :post
+
+    after_save :update_post_likes_count
+
+    private
+
+    def update_post_likes_count
+        post_likes_count = Comment.where(post_id: post_id).count
+        post.update(likesCounter: post_likes_count)
```
Comment (bienvenuushindi) [inline app/models/likes]: :red_circle: [REQUIRED] Please ensure that you adhere to Ruby's best practices for variable naming. The naming convention for variables is` snake_case` for methods and variables. So `likesCounter` should be `likes_counter`

Comment (bienvenuushindi) [inline app/models/likes]: :red_circle: [REQUIRED] Please ensure that you adhere to Ruby's best practices for variable naming. The naming convention for variables is` snake_case` for methods and variables. So `likesCounter` should be `likes_counter`

File: app/models/likes [hunk 6]
```diff
@@ -0,0 +1,13 @@
...
+    end
```
Comment (bienvenuushindi) [inline app/models/likes]: :orange_circle: [OPTIONAL] Below is a more concise and clear way to achieve the same result as the previous code:

def update_post_likes_count
post.increment!(:likesCounter)
end

File: app/models/comment.rb [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+  belongs_to :user
+  belongs_to :post
+
+  after_save :update_post_comments_count
+
+  private
+
+  def update_post_comments_count
+    post_comments_count = Comment.where(post_id:).count
+    post.update(commentsCounter: post_comments_count)
```
Comment (bienvenuushindi) [inline app/models/comment.rb]: :red_circle: [REQUIRED] Please ensure that you adhere to Ruby's best practices for variable naming. The naming convention for variables is` snake_case` for methods and variables. So `commentsCounter` should be `comments_counter`

File: app/models/comment.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+  end
```
Comment (bienvenuushindi) [inline app/models/comment.rb]: :orange_circle: [OPTIONAL] Below is a more concise and clear way to achieve the same result as the previous code:
```

def update_post_comments_count
    post.increment!(:commentsCounter)
end
```

**Review by bienvenuushindi — CHANGES_REQUESTED** [review]
Hi @Elerqsousy  , 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!


## Highlights
May I highlight the following:

- The repo looks professional.:heavy_check_mark:
- There are no linter errors:heavy_check_mark:
- You have descriptive PR:heavy_check_mark:
- Your project looks good:heavy_check_mark:

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Check the comments under the review._

_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/post.rb (hunk 3 above)
Comment (Elerqsousy) [reply app/models/post.rb]: Thank you for the amazing documentation of comments.
I really appreciate your support.
I would like to discuss this part if you don't mind. As I am using `after_save` to call this method on each instance update. Upon deletion it will be called as well and will leave the counter without updating the right count. on another note, on updating any post, it will increment the post counter by one, which shouldn't be happening.
If I may will not include this in my code.

Comment (Elerqsousy) [reply app/models/post.rb]: @bienvenuushindi kindly advise.
Thank you in advance

Comment (bienvenuushindi) [reply app/models/post.rb]: @Elerqsousy I appreciate your feedback :+1:

Comment (bienvenuushindi) [reply app/models/post.rb]: @Elerqsousy You are absolutely right :bulb:  :+1: concerning the issue :bug: of the counter not getting updated after a deletion. To address this, it's a good practice to implement a dedicated method to handle the deletion operation, and then you can invoke this method using the after_destroy filter. 

Here's how you can implement it:

```
after_destroy :update_user_posts_counter_on_destroy

def update_user_posts_counter_on_destroy
  author.decrement!(:postsCounter)
end

```

Comment (bienvenuushindi) [reply app/models/post.rb]: @Elerqsousy  :bulb: It's also important to note that, in this case, the preferable filter should be `after_create` rather than `after_save`. `after_save` executes actions after a record is saved, regardless of whether it's a new record (create) or an existing record, whereas `after_create` specifically triggers an action after a new record is successfully created, excluding updates to existing records.

Comment (bienvenuushindi) [reply app/models/post.rb]: In simpler terms:

after_save can run multiple times for the same record if it's updated multiple times.

after_create runs only once, and it's specifically triggered after a new record is successfully created. It doesn't run during updates to existing records.

**Review by BertrandConxy — APPROVED** [review]
Hi @Elerqsousy, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights 💪 
- No linter errors
- project requirements are met
- project works fine

## Optional suggestions
  - [x] Did you see that your likes file don't have `.rb` file format. Saw that It has been mentioned but PLEASE make sure your likes file has the file format and become 👉🏾  `likes.rb`

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @BertrandConxy in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
